utilities/transactions/transaction_base.cc

```diff
@@ -21,6 +21,43 @@
 
 namespace ROCKSDB_NAMESPACE {
 
+Status Transaction::CommitAndCreateSnapshot(
+    std::shared_ptr<TransactionNotifier> notifier, TxnTimestamp ts,
+    std::shared_ptr<const Snapshot>* snapshot) {
+  TxnTimestamp commit_ts = GetCommitTimestamp();
+  if (commit_ts == kMaxTxnTimestamp) {
+    if (ts == kMaxTxnTimestamp) {
+      return Status::InvalidArgument("Commit timestamp unset");
+    } else {
+      const Status s = SetCommitTimestamp(ts);
+      if (!s.ok()) {
+        return s;
+      }
+    }
+  } else if (ts != kMaxTxnTimestamp) {
+    if (ts != commit_ts) {
+      // For now we treat this as error.
+      return Status::InvalidArgument("Different commit ts specified");
+    }
+  }
+  // Do not try to dereference prev_snapshot because it can be destroyed.
+  const Snapshot* const prev_snapshot = GetSnapshot();
+  SetSnapshotOnNextOperation(notifier);
+  Status s = Commit();
+  if (!s.ok()) {
+    return s;
+  }
+  std::shared_ptr<const Snapshot> new_snapshot = GetTimestampedSnapshot();
+  if (new_snapshot.get() == prev_snapshot) {
+    return Status::NotSupported();
+  }
+
+  if (snapshot) {
```

Comment:
How likely we return snapshot==null? It feels to me that it is unlikely. If that is the case, would it be better to turn it into a status code that is not OK?

Reply:
I think this is one way. I just feel like the API is not obvious what a null return means. I guess it means the timestamp is already used by a previous snapshot?

Reply:
oh I think I misunderstood. You meant that we should return a non-ok status if we fail to create a timestamped snapshot internally. It's not related to the specified argument. If so, I think we can return `kInComplete` or `kAborted`.

Reply:
> I guess it means the timestamp is already used by a previous snapshot?

Currently, for write-unprepared/write-prepared, it can be null because no timestamped snapshot is created. It just means unsupported, and we return that status code already.
For write-committed, if the timestamp is specified, and commit is successful, we should always be able to create a timestamped snapshot and return it.
Maybe we should emphasize the following
```cpp
// It is also assumed that application will
ensure any two snapshots with timestamps should satisfy the following:
snapshot1.seq < snapshot2.seq iff. snapshot1.ts < snapshot2.ts
```

Reply:
I will think more about this. We can perform some checks and return non-ok if the above is violated.